Level 3

VitalyTikhonov opened the pull request.

VitalyTikhonov committed:
F002 Configured Webpack

VitalyTikhonov committed:
F003 + .nojekyll

VitalyTikhonov committed:
F004 Set Webpack up for two pages

VitalyTikhonov committed:
F005 Made HTML Mobile_Main_Results_Not_Logged_In

VitalyTikhonov committed:
F006 Bulk-added css files

VitalyTikhonov committed:
F007 Added imports to CSS files

VitalyTikhonov committed:
F008 Set 320 width, added root__child-wide class

VitalyTikhonov committed:
F009 Making search-panel, header
Header: fixed position, issue: layout valid in dev tools devices mode only

VitalyTikhonov committed:
F010 Switched root child width and padding to rems

VitalyTikhonov committed:
F011 Search panel done

VitalyTikhonov committed:
F012 Started styling the article block + more
Added paddings to the article block, added the section-headline class, reworked card__breadcrumb classes.

VitalyTikhonov committed:
F013 Fixed some path errors

VitalyTikhonov committed:
F014 Switching from html-rem to root-em
Removed rems, moved the base font from html to root

VitalyTikhonov committed:
F015 Cancelled the em-rem attempts, plain pixels

VitalyTikhonov committed:
F016 Making the card: 80%

VitalyTikhonov committed:
F017 Card done

VitalyTikhonov committed:
F018 Card container as a grid

VitalyTikhonov committed:
F019 More button and gen button class made

VitalyTikhonov committed:
F020 User profile made

VitalyTikhonov committed:
F021 Making the footer

VitalyTikhonov committed:
F022 Footer made

VitalyTikhonov committed:
F023 Main menu positioned; format pending

VitalyTikhonov committed:
F024 Main menu made

VitalyTikhonov committed:
F025 Added the hidden class for display-none

VitalyTikhonov committed:
F026 Mob n/li: made saved-news-intro

VitalyTikhonov committed:
F027 The Webpack Two pages Battle going on...

VitalyTikhonov committed:
F028 Webpack Two Pages working version 2
For parallel html positioning; suboptimal still working js- and css-links in html

VitalyTikhonov committed:
F029 Webpack Two Pages Non-Parallel: DefinePlugin
A failed attempt to solve the image link issue in non-parallel HTML files using the DefinePlugin and the ugly instructions at the Webpack site.

VitalyTikhonov committed:
F030 Webpack 2P !par: non-BEM, prod fails, dev ok

VitalyTikhonov committed:
F031 Webpack 2P PAR: partially BEM, b&d working
Webpack two pages: parallel HTML placement, partially per BEM, both build and dev working, second page available at /saved-news.html.

VitalyTikhonov committed:
F032 Second page made but the menu

VitalyTikhonov committed:
F033 Made 3step menu toggle scheme for both mob pages

VitalyTikhonov committed:
F034 Mob n/li login popup complete but input__errors

VitalyTikhonov committed:
F035 Mob n/li login popup complete

VitalyTikhonov committed:
F036 Optimized image files

VitalyTikhonov committed:
F037 Added GH-Pages to Readme

VitalyTikhonov committed:
F038 Edited Readme

VitalyTikhonov committed:
F039 A minor change: obsolete strings

VitalyTikhonov committed:
F040 Tablet: header pending, search-panel done

VitalyTikhonov committed:
F041 Tablet: header & menu & search-panel done

VitalyTikhonov committed:
F042 Tablet: header search-panel 320-768-adapted

VitalyTikhonov committed:
F043 Tablet: article-container 320-768-adapted

VitalyTikhonov committed:
F044 Tablet: article-block 320-768-adapted

VitalyTikhonov committed:
F045 TABLET 320-768-adapted

VitalyTikhonov committed:
F046 Touchpad issue overcome; Tb ix footer fine-tuned

VitalyTikhonov committed:
F047 Tb html comparison, adapting Sn

VitalyTikhonov committed:
F048 Header formatting conflicts discovered

VitalyTikhonov committed:
F049 The header scheme rework: mostly done

VitalyTikhonov committed:
F050 Header scheme rework: box-shadow fixed

VitalyTikhonov committed:
F051 The button collapse (failed build)

VitalyTikhonov committed:
F052 Button Set Development

VitalyTikhonov committed:
F051 Button Set Development v0.1

VitalyTikhonov committed:
F052 Button Set Development v0.2

VitalyTikhonov committed:
F053 Added the button set to blocks

VitalyTikhonov committed:
F054 Restoring webpak.config

VitalyTikhonov committed:
F055 Restoring webpak.config

VitalyTikhonov committed:
F056 WEBPACK 2P PAR, js&css in pages[name]

VitalyTikhonov committed:
F057 Button Set Development v0.3

VitalyTikhonov committed:
F058 Implementing the button set and some o10n

VitalyTikhonov committed:
F059 Improved the visibility classes

VitalyTikhonov committed:
F060 Finished Mob-Tab debugging

VitalyTikhonov committed:
F061 A typo in a class name fixed

VitalyTikhonov committed:
F062 Card breadcrumb tooltip set

VitalyTikhonov committed:
F063 Desktop primary edit

VitalyTikhonov committed:
F064 Desktop-adapted

VitalyTikhonov committed:
F065 Popups finished

VitalyTikhonov committed:
F066 Reworked and completed the links

VitalyTikhonov committed:
F067 W3C Check Adjustments: button beyond ul, ids in forms, typos

VitalyTikhonov committed:
F068 W3C Check Adjustments
Versions, input focus and validation

VitalyTikhonov committed:
F069 Corrections per review: backup fonts, a required attr

VitalyTikhonov committed:
F070 WebPack: switch to non-parallel html placing
Src: BEM-like, Dist: technology-based + saved-news/

VitalyTikhonov committed:
F071 Making BaseComponent and Popup

VitalyTikhonov committed:
F072 BaseComp listener method, Popup open&close

VitalyTikhonov committed:
F073 Popup esc&click-away; making Signup form
About to move popup instantiating to a callback

VitalyTikhonov committed:
F074 FormPopup-SignupForm attempt suspended

VitalyTikhonov committed:
F075 Popup+Form>Header attmpt suspended

VitalyTikhonov committed:
F076 Made form show: Form&Popup instantiation on load

VitalyTikhonov committed:
F077 Vaking the Api class
About to replace the Form callback with a direct instance.

VitalyTikhonov committed:
F078 Replace Form callback with direct instance
To keep index file structure while chaining MainApi - Form - Popup, as well as for expected use of other Form methods in Popup.

VitalyTikhonov committed:
F079 Making Form & Api, refactored constants, etc.

VitalyTikhonov committed:
F080 Set up Backend error message reading
(Backend: added CORS)

VitalyTikhonov committed:
F080 Added gen err mes show & form closing

VitalyTikhonov committed:
F081 Moving common op6s to BaseComponent

VitalyTikhonov committed:
F082 Reworked popup open with renewed BaseComponent

VitalyTikhonov committed:
F083 Implemented formDismissal custom event
To eliminate the need of method interchange between the Popup and Form classes

VitalyTikhonov committed:
F084 Previous commit feature: enabled listener removal

VitalyTikhonov committed:
F085 The strange useCapture omission fixed

VitalyTikhonov committed:
F086 The auth-button remaining in focus bug fixed

VitalyTikhonov committed:
F087 Making confirmation popup

VitalyTikhonov committed:
F088 Confirmation popup opens based on a custom event

VitalyTikhonov committed:
F089 Change of auth popups implemented

VitalyTikhonov committed:
F090 Made and debugging login request

VitalyTikhonov committed:
F091 Login request completed

VitalyTikhonov committed:
F092 Making tne NewsApi class

VitalyTikhonov committed:
F093 Refactoring the Form class

VitalyTikhonov committed:
F093 Refactoring the Form class

SmilingJey commented:
Здравствуйте, работа отклонена от проверки, при открытии падают ошибки, как при локальной сборке, так и на gh-pages
Так же хочу обратить внимание, что проект должен быть настроен на работу с удаленным сервером, сейчас в коде `const API_URL = `http://localhost:${PORT}`;`

![Screenshot_1](https://user-images.githubusercontent.com/9317133/97278799-40943280-184b-11eb-874f-80de59defc29.png)
![Screenshot_2](https://user-images.githubusercontent.com/9317133/97278811-42f68c80-184b-11eb-84ad-81360d4afb24.png)

VitalyTikhonov committed:
Merge branch 'level-3' of github.com:VitalyTikhonov/News_Explorer_Frontend into level-3
Gt has just stolen my commit that should be:
F094 NewsApi working; NSForm as BC's, not Form's successor

VitalyTikhonov committed:
F095 Removed cancelled UserDataForm successor of Form

VitalyTikhonov committed:
F096 Debugged issues that resulted in recent class changes

VitalyTikhonov committed:
F096 Minor adjustments

VitalyTikhonov committed:
F097 Added the AccessControl class

VitalyTikhonov committed:
F098 Minor adjustments

VitalyTikhonov committed:
F099 Corrected page addresses in htmls

VitalyTikhonov committed:
F100 Making class setting methods: diff vars, nonfunc build

VitalyTikhonov committed:
F101 Making class setting methods

VitalyTikhonov committed:
F102 Class setting methods & Menu open/close made

VitalyTikhonov committed:
F103 Logout made

VitalyTikhonov committed:
F104 Implementing Article-Block: Preloader

VitalyTikhonov committed:
F105 Preloader and No-news bumper made

VitalyTikhonov committed:
F106 Card render: initial implementation

VitalyTikhonov committed:
F107 Card render: date formating

VitalyTikhonov committed:
F108 Card render: sizing, image & source position, link

VitalyTikhonov committed:
F109 Card save button: wrong configuration

VitalyTikhonov committed:
F110 Card save button: button activation made

VitalyTikhonov committed:
F111 Removed obsolete lines

VitalyTikhonov committed:
F112 Remade article block to render from scratch, debugging news search

VitalyTikhonov committed:
F113 Debugged article block content change & card save button enablement issues

VitalyTikhonov committed:
F114 Implemented news request error message in popup

VitalyTikhonov committed:
F115 Made article default image

VitalyTikhonov committed:
F116 Made article save & delete

VitalyTikhonov committed:
F117 Made tooltip text change

VitalyTikhonov committed:
F118 Tooltip text change on save/delete; or-signin link margin

VitalyTikhonov committed:
F119 More button

VitalyTikhonov committed:
F120 Failed attempt to make centralized event handler removal

VitalyTikhonov committed:
Revert "F119 More button"

This reverts commit 3a62d51451f2fb71544a1c074cb84c8c3c0f78be.

VitalyTikhonov committed:
Revert "F120 Failed attempt to make centralized event handler removal"

This reverts commit 5f8e1bb65b8fe58b820696fa8ae7733173fe4190.

VitalyTikhonov committed:
Revert "Revert "F119 More button""

This reverts commit 85cc4204f524c43e4cbcb92bce26838abb5815db.

VitalyTikhonov committed:
F121 Started second page implementation

VitalyTikhonov committed:
F122 Page 2: menu open&close

VitalyTikhonov committed:
F123 Page2: redirects for non-authorized users

VitalyTikhonov committed:
F124 Page2: basic preps for article render

VitalyTikhonov committed:
F125 Debugged page 1 card portion render: more button, save button state

VitalyTikhonov committed:
F126 Page 2: made card render and deletion

VitalyTikhonov committed:
F127 Preps to display username in logout button text

VitalyTikhonov committed:
F128 Made display of username in logout button

VitalyTikhonov committed:
F129 Deploy: paths, tree, page 2 card render
Corrected the path for production to api..., changed the initial file tree so that the image paths and the redirect in page 2 worked in the output, debugged card fields render

VitalyTikhonov committed:
F130 Reinstated deep-nested page2 output, optimized images

VitalyTikhonov committed:
F131 Making saved news intro: spare revert point

VitalyTikhonov committed:
F132 SavedNewsIntro: section render & headline

VitalyTikhonov committed:
F133 SavedNewsIntro: name, number of articles and keyword summary

VitalyTikhonov committed:
F134 SavedNewsIntro: making the updating method

VitalyTikhonov commented:
http://vitaliytikhonov.ru/webdev/projects/news_explorer/
Смотрите, пожалуйста, предпоследний коммит - F133. Последний еще сырой.

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--8.0--> Настроена работа API для поиска новостей (https://newsapi.org/) и работы с сервером. Любые другие внешние API запрещены.
- Асинхронные запросы к API:
  - [x] <!--6.0--> запросы можно осуществлять через Fetch API или XMLHttpRequest; сторонние библиотеки запрещены — такие как axios или jQuery;
  - [x] <!--5.5--> для работы с API создан отдельный класс;
  - [x] <!--5.5--> цепочка обработки промисов завершается блоком catch;
  - [x] <!--6.0--> первый обработчик then возрващает res.json;
  - [x] <!--6.0--> манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
- Именование:
  - [x] <!--2.5--> camelCase для параметров, методов и переменных;
  - [x] <!--1.5--> существительные для имён переменных;
  - [x] <!--1.5--> существительные во множественном числе для NodeList;
  - [x] <!--2.5--> существительные с заглавной буквы для имён классов;
  - [x] <!--2.5--> const для неизменяемыx напрямую переменных;
  - [x] <!--1.5--> соответствующее содержимому имя класса;
  - [x] <!--1.5--> отсутсвует транслит и неуместные сокращения.
- [x] <!--6.0--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--3.5--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»
- [x] <!--8.0--> JWT сохраняется в браузере.
- [x] <!--6.0--> Код JS-модулей инкапсулирован.
- [ ] <!--6.0--> Поля форм входа и регистрации валидируются на клиенте.
**Комментарий: запрос выполняется даже при пустых полях ввода**

### Хорошие практики
- Код объектно-ориентирован:
  - [x] <!--0.7--> использованы ES6-классы;
  - [x] <!--1.1--> внутри классов нет создания экземпляров других классов;
  - [x] <!--0.7--> соблюдён принцип DRY (Don't Repeat Yourself);
  - [x] <!--1.2--> соблюдён принцип единственной ответственности, то есть каждый класс полностью отвечает за свою задачу, и никакой другой класс к решению этой задачи не относится;
  - [x] <!--1.2--> соблюдён принцип открытости-закрытости: для добавления функциональности классы можно расширять, не модифицируя их;
  - [x] <!--0.7--> у классов есть приватные поля, то есть свойства и методы, которые начинаются с нижнего подчёркивания и не используются нигде вне класса;
- Работа с DOM-деревом для вывода результатов:
  - [x] <!--1.2--> cайт защищён от XSS: нельзя выполнять скрипты через пользовательский ввод;
  - [x] <!--0.7--> нет потенциальных утечек памяти: обработчики созданы и удалены в нужных местах;
  - [x] <!--0.7--> NodeList перебирается с использованием методов для работы с массивами и псевдомассивами;
  - [x] <!--0.7--> множественные операции c DOM-элементами производятся прежде, чем добавлять блок в разметку;
  - [x] <!--0.7--> DOM-элементы не прерисовываются без необходимости;
  - [x] <!--1.2--> innerHTML запрещён — вместо него применяются более мягкие методы;
  - [x] <!--0.7--> поиск по DOM-дереву выполняется минимально возможное количество раз, то есть поиск одного элемента не выполняется дважды;
- Обработка ошибок API:
  - [x] <!--1.2--> пользователь получает сообщение в случае ошибки;
  - [ ] <!--0.7--> поля формы заблокированы во время отправки запросов;
  **Комментарий: на время выполнения запроса нужно блокировать поля ввода и кнопку отправки формы**
- [x] <!--1.2--> Разные логические части приложения описаны в разных файлах и разнесены в разные папки. Интерфейсные классы, классы API, вспомогательные функции и константы отделены друг от друга.
- [x] <!--0.7--> При клике на «Выйти» страница автоматически меняет состояние и выглядит так, как у пользователя, не выполнившего вход в систему.

### Рекомендации
- [x] <!--0.6--> прелоудер крутится при ожидании ответа от сервера
- [x] <!--0.3--> `event.target` и `this` применяются правильно;
- Обработка и валидация форм:
  - [ ] <!--0.3--> текст ошибок кастомный, а не системный;
  - [ ] <!--0.3--> валидация реализована и представляет собой отдельный метод для обработки форм
- Отсутствует «мусор» в коде:
  - [x] <!--0.3--> по имени колбэка слушателя понятно его назначение;
  - [x] <!--0.3--> нет «магических значений», все значения записаны в переменные;
  - [x] <!--0.3--> константы описаны в той области видимости, в которой объявлена функция, и внутри неё значения констант берутся из замыкания;
  - [x] <!--0.3--> если с API можно работать несколькими способами, то выбран один способ, который применяется везде в проекте. Например, поиск отдельных элементов осуществляется только методом `querySelector`
- [x] <!--0.3--> Любой попап закрывается не только при клике на иконку крестика, но и при клике в любое место вне попапа и при нажатии клавиши Esc.
- [x] <!--0.4--> После успешной регистрации пользователю показывается попап с сообщением: «Пользователь успешно зарегистрирован!» и предложением войти в систему
- [x] <!--0.4--> Если пользователь не вошёл в систему, то переход на страницу с сохранёнными новостями по прямой ссылке осуществляет редирект на главную.
- [ ] <!--0.4--> Поля форм входа и регистрации валидируются на клиенте. Сообщения об ошибках кастомные, а не стандартные. Кастомные тексты ошибок есть в макете.
- [x] <!--0.3--> Создан базовый компонент BaseComponent. Другие компоненты интерфейса наследуют у него.

## Количество баллов: 92

## Доп. Комментарии:
- если используются модули es6, то оборачивать их содержимое в IIFE не требуется

## Хорошая работа, поздравляю с завершением дипломного проекта!  Однако, если планируете использовать проект в портфолио советую ещё поработать над ним - сделать валидацию форм, заголовок при скроле сейчас начинается сливаться с белым фоном, у попапа не отобразилась иконка закрытия, на странице сохраненных ссылка главная ведет не туда и проверить весь функционал и отображение сайта.

### Несколько советов для дальнейшего развития:
- свертайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- изучить фреймворк - React или Vue.js, хотя бы на базовом уровне
- ходить на собеседования, опыт прохождения собеседований не менее важен знаний теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушаете подкаксты:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

VitalyTikhonov commented:
Спасибо!

VitalyTikhonov committed:
F135 Refactoring: images on Page 2

VitalyTikhonov committed:
F136 Ref. Page2: links, file structure

VitalyTikhonov committed:
F137 Ref. Image default url

VitalyTikhonov committed:
F138 Ref. Not-found image url

VitalyTikhonov committed:
F139 Ref. PublicPath dev/prod for good file paths

VitalyTikhonov committed:
F140 Ref. PublicPatch for good dev/prod file paths

VitalyTikhonov committed:
F141 Ref. Form class & other preps for form validation

VitalyTikhonov committed:
F142 Ref. Making form validation

VitalyTikhonov committed:
F143 Ref. Dialog forms validation
